enhancements/microshift/audit-log-configuration-options.md

```diff
@@ -0,0 +1,249 @@
+
+Both OpenShift and kube-apiserver allow defining the depth of collected API request (metadata, request, response).  OpenShift extends the definition of these profiles to filter messages based on the request's verb (see table above). It will therefore be necessary to implement logic in MicroShift to support these extended definitions. This is possible and can be done via the Policy.audit.k8s.io/v1 object. Each kube-apiserver level can be extended to filter per verb, per API resource.  Thus, MicroShift will have to know how to map a given profile to a backend Policy object such that the message scope, verb and resource exactly represent the OpenShift profile.
+
+The Kubernetes Policy spec can be found [here](https://kubernetes.io/docs/reference/config-api/apiserver-audit.v1/#audit-k8s-io-v1-Policy).
+
+A simple example of the OpenShift "Default" profile could be represented in a Policy object as:
+
+```yaml
+apiVersion: audit.k8s.io/v1
+kind: Policy
+metadata:
+  name: Default
+omitStages:
+- "RequestReceived"
+rules:
+- level: Metadata
+    verbs:
+    - create
+    - update
+    - patch
+    - delete
+    - deletecollection
+    - get
+    - list
+    - create
+    - update
+    - patch
+    resources:
+    - group: "*"
+      resources: "*"
+```
+
+Note that "omitStages" is determines the point at which a message is logged. For more on stages, see [here](https://kubernetes.io/docs/tasks/debug/debug-cluster/audit/). Omitting the "RequestRecieved" stage is preferable to cut down on log spam but may be useful for debugging. 
+
+### Risks and Mitigations
+
+- Exceeding disk capacity: Microshift targets small form-factor devices with limit on-board storage. If the product of `maxFileSize` and `maxFiles` equals a size larger than the available storage, the apiserver risks destabilizing the system by maxing out the consumed storage.  This can be mitigated via documentation which recommends the user understand their storage limitations when setting these values. Logging a warning during startup when the apiserver values exceed the available storage at the logging path. It is impossible to predict which component of MicroShift or the host system will fail when the disk is out of space, thus logging a calculated overage can provide an important clue.
+- Lost Log Data: The apiserver culls log files given a certain size or age. If users do not take care to back up logs at a rate greater than the rate at which the kube-apiserver culls the files, they risk losing log data. This can be mitigated with examples of log-forwarding provided in documentation, example manifests, or both. Alternatively, it may be necessary to consider deploying the openshift-cluster-logging-operator to provide supportable log forwarding features.
```

Comment:
Is there a way for MicroShift to detect when log rotation / truncation occurs and give users a hook to run their handler, i.e. script that copies data elsewhere?

Reply:
They could even use something like inotify - program that detects file/dir operation like read, write, chmod, etc. That could be their hook to create script around it

Reply:
Maybe, but I think it's at this point log-forwarding should be the next step.